core_types/src/time.rs

```diff
@@ -42,35 +42,135 @@ impl From<usize> for Timeout {
     }
 }
 
-/// This struct represents datetime data stored as a string
-/// in the ISO 8601 format.
+/// This struct represents datetime data stored as a string in the ISO 8601 and RFC 3339 (more
+/// restrictive) format.
+///
 /// More info on the relevant [wikipedia article](https://en.wikipedia.org/wiki/ISO_8601).
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct Iso8601(String);
 
+/// A static string is considered an infallible conversion; also unchecked infallible String conversion.
+///
+/// Since we receive Iso8601 from many remote, untrusted sources, we don't want to always force
+/// checking at initial creation.  However, later on, we need to validate the timestamp, and we may
+/// want to create a validated Iso8601 that we know will work in comparisons, etc.  For that, we use
+/// the TryFrom method.
+///
+/// In the future, any invalid static-&str errors could (should?) produce a `panic!`.  This is
+/// reasonable, as it would indicate an error in the code of the application, not in the logic.
 impl From<&'static str> for Iso8601 {
     fn from(s: &str) -> Iso8601 {
         Iso8601(s.to_owned())
     }
 }
 
-pub fn test_iso_8601() -> Iso8601 {
-    Iso8601::from("2018-10-11T03:23:38+00:00")
+impl From<String> for Iso8601 {
+    fn from(s: String) -> Iso8601 {
```

Comment:
why not put the validity check in the construction of the `Iso8601`?

Reply:
I experimented with that a bit. We could, but that was a more drastic step than I wanted to embark on at this time -- I'm trying to get the Holofuel transactor working!  I think a reasonable next step would be to add an `is_valid` method, though, and look at what fails when we invoke that in the infallible `From` trait implementations, which would have to panic, which I don't like.  A lot of tests failed, to start with.  This is a first step...

My feeling is that allowing incoming data with whatever timestamps is OK, as long as you don't have to compare them -- then be more strict.  It's now easy to filter out and report/reject items with bad Iso8601s and produce a less catastrophic HolochainError, for example.

Reply:
Since we pass a lot of Iso8601 back and forth across external APIs, I think causing these APIs to fail in deserialization would be worse than validating the timestamps in the target function, according to the rules of the application.

Reply:
@pjkundert `TryFrom`?

Reply:
@pjkundert look at our [jsonstring](https://github.com/holochain/holochain-rust/blob/develop/doc/holochain_101/src/json_string.md) docs for how to do this...

Reply:
OK, I've implemented JSON serialization via both serde::Serialize/Deserialize, and via TryFrom<JsonString> for Iso8601 and From<Iso8601> for JsonString.  We do not want to fail invalid Iso8601 values in transit, I believe, right now. Too many tests to fix, for one thing.

I've added a simple check for validity via `Iso8601.validated().is_ok()`, and optimized the whole process to only attempt conversion once on creation.  We round-trip the canonicalized, valid ISO 8601 timestamp if possible, or retain the original invalid one -- JSON round-tripping still only involves one String, but it'll be the canonicalized version, if possible.